### DISCUSSION TITLE
Pinball running on T-Display-S3 based on the Ten Minute Physics Tutorial "How to write a pinball simulation."

### DISCUSSION THREAD

**Discussion by russhughes** [micropython/micropython#9826] [Show and tell]
This was my latest weekend project, a pinball game for the T-Display-S3 board based on the Ten Minute Physics Tutorial "How to write a pinball simulation."

- [Game Play Video](https://www.youtube.com/watch?v=y0B3i_UmEU8)
- [MicroPython Source](https://github.com/russhughes/st7789s3_mpy/blob/main/examples/pinball.py) (Requires [tft_config.py](https://github.com/russhughes/st7789s3_mpy/blob/main/examples/tft_config.py) module)
- [Firmware Repo with display driver](https://github.com/russhughes/st7789s3_mpy)
- [Ten Minute Physics Tutorial "How to write a pinball simulation."](https://youtu.be/NhVUCsXp-Uo)

<img width="202" alt="Screen Shot 2022-11-02 at 11 39 52 AM" src="https://user-images.githubusercontent.com/23587382/199574562-88177195-5274-4672-b543-43eed2698401.png">

**Comment by rkompass** [comment]
Nice and impressing. Also good to see, that ordinary drawing is fast enough to make the animation smooth.
Did you encounter any difficulties in the process of development, which are worth sharing?

**Reply by rkompass** [reply]
This was a superficial question. I had a look at the sources now and see that the display driver is coded in C and you added and modified a lot to it. There is no framebuffer involved and the drawing of hlines and vlines is done in the display hardware ?!?
The driver is incorporated into a ESP32 firmware.
Questions now arising in my mind are:

- Could it be incorporated into pyboard firmware as well and into RPi Pico firmware?
- Would it be possible to generate a loadable module from it?
- Which other display hardware is similarly usable for this approach?
- Why did you decide against using the framebuffer which is already there in uPy? Memory consumption, SPI speed, ....?
- Was the usage of lvgl considered as an option for the graphics?

Congratulations on your achievement.

**Reply by russhughes** [reply]
Thank you. I wrote a pinball game years ago but my approach was very naïve and in the end, it was not fun to play. I ran across the Ten Minute Physics Tutorials and it made sense to me.
 
`Could it be incorporated into pyboard firmware as well and into RPi Pico firmware?
`

Yes, this version of the driver is specific for the T-Display-S3's 8-bit parallel display. It uses the ESP32-S3's dedicated GPIO (CPU accessible 8-bit port) for speed. Replacing the dedicated GPIO routines with generic GPIO routines would allow it to run on other processors. The version from [https://github.com/russhughes/st7789_mpy]( https://github.com/russhughes/st7789_mpy) is for SPI-based displays. It is more generic, with firmware for ESP32, RP2040, and PYBV11 devices.
 
`Would it be possible to generate a loadable module from it?`

The last time I checked, it looked challenging to do. Native modules have limited access to MicroPython internal library functions. Loadable modules are more suited to speeding up CPU-intensive operations than doing hardware I/O. Using a native module to write to a framebuffer looks doable.
 
`Which other display hardware is similarly usable for this approach?`

Many displays like the ili93xx, st77xx, gc9a01, and other display controllers use a similar protocol with some differences in the display initialization and other parameters.   
 
`Why did you decide against using the framebuffer, which is already there in uPy? Memory consumption, SPI speed, ....?`
 
Memory consumption mainly. I modified [devbis]( https://github.com/devbis/st7789_mpy) driver to add text capability for my [TurtlePlotBot]( https://github.com/russhughes/turtleplotbot3) project. The original processor did not have much memory. It’s grown from there, mainly as a learning experience. 
 
`Was the usage of lvgl considered as an option for the graphics?`
 
LGVL is excellent for creating nice user interfaces but is very large when added to MicroPython.  I wrote a pull request that was accepted and merged into LVGL to add support for the st7789 displays that I like to use, so I will probably use it in an upcoming project.
